File: src/pset6/credit/credit.py
from cs50 import get_int

AMERICAN_EXPRESS = 34
AMERICAN_EXPRESS_2 = 37
AMERICAN_EXPRESS_LEN = 15
MASTER_CARD = 51
MASTER_CARD_2 = 52
MASTER_CARD_3 = 53
MASTER_CARD_4 = 54
MASTER_CARD_5 = 55
MASTER_CARD_LEN = 16
VISA = 4
VISA_CARD_LEN = 13
VISA_CARD_LEN_2 = 16
MAX_CARD_LEN = 16
MIN_CARD_LEN = 13


def main():
    # Get arguments
    card_number = get_int("Number: ")
    card_number_str = str(card_number)
    card_number_str_len = len(card_number_str)

    # Check length
    if card_number_str_len < MIN_CARD_LEN and card_number_str_len > MAX_CARD_LEN:
        print("INVALID")
        return 1

    # Holds the total value
    total = 0

    # Multiply all digits by 2, starting from the second last digit,
    # and add the digits of the resulting numbers
    for i in range(card_number_str_len - 2, -1, -2):
        digit = int(card_number_str[i])
        twice_of_digit = digit * 2
        twice_of_digit_str = str(twice_of_digit)
        twice_of_digit_str_len = len(twice_of_digit_str)
        for j in range(twice_of_digit_str_len):
            twice_of_digit_digit = int(twice_of_digit_str[j])
            total = total + twice_of_digit_digit

    # Scroll the numbers not multiplied by two
    for i in range(card_number_str_len - 1, -1, -2):
        digit = int(card_number_str[i])
        total = total + digit

    # Get total str
    total_str = str(total)

    # Return 'INVALID' if the last digit of the total is not equal '0'
    if total_str[-1:] != '0':
        print("INVALID")
        return 1

    # Get tast digits
    first_digit = int(card_number_str[0:1])
    first_two_digit = int(card_number_str[0:2])

    # Check card types
    if first_digit == VISA and card_number_str_len >= VISA_CARD_LEN and card_number_str_len <= VISA_CARD_LEN_2:
        print("VISA")
    elif (first_two_digit == MASTER_CARD or first_two_digit == MASTER_CARD_2 or first_two_digit == MASTER_CARD_3 or first_two_digit == MASTER_CARD_4 or first_two_digit == MASTER_CARD_5) and card_number_str_len == MASTER_CARD_LEN:
        print("MASTERCARD")
    elif (first_two_digit == AMERICAN_EXPRESS or first_two_digit == AMERICAN_EXPRESS_2) and card_number_str_len == AMERICAN_EXPRESS_LEN:
        print("AMEX")
    else:
        print("INVALID")

    # End of main function
    return 0

# Call the main function
if __name__ == "__main__":
    main()
    pass

File: src/pset1/credit/credit.c
#include <stdio.h>
#include <math.h>

#include "../../../lib/cs50.h"

typedef char byte;

const byte AMERICAN_EXPRESS = 34;
const byte AMERICAN_EXPRESS_2 = 37;
const byte AMERICAN_EXPRESS_LEN = 15;

const byte MASTER_CARD = 51;
const byte MASTER_CARD_2 = 52;
const byte MASTER_CARD_3 = 53;
const byte MASTER_CARD_4 = 54;
const byte MASTER_CARD_5 = 55;
const byte MASTER_CARD_LEN = 16;

const byte VISA = 4;
const byte VISA_CARD_LEN = 13;
const byte VISA_CARD_LEN_2 = 16;

const byte MAX_CARD_LEN = 16;
const byte MIN_CARD_LEN = 13;

byte get_digit_count(long l)
{
    int count = 1;
    while (l /= 10)
        count++;
    return count;
}

long get_digit(long l, long m)
{
    l %= m;
    if (m != 10)
    {
        m /= 10;
        l -= l % m;
    }
    return l != 0 ? (l > 9 ? l / m : l) : 0;
}

int main(void)
{
    // Get card information
    long credit_card_number = get_long("Number: ");
    byte credit_card_digit_count = get_digit_count(credit_card_number);

    if (credit_card_digit_count < MIN_CARD_LEN || credit_card_digit_count > MAX_CARD_LEN)
    {
        printf("INVALID\n");
        return 0;
    }

    long total = 0, grand_total = 0;
    long m = 10;
    byte double_coefficient = credit_card_digit_count % 2 == 0 ? 1 : 0;
    for (int i = credit_card_digit_count; i > 0; i--)
    {
        long digit = get_digit(credit_card_number, m);
        if ((i + double_coefficient) % 2 == 0)
        {
            long twice_of_digit = digit * 2;
            if (twice_of_digit < 10)
            {
                total += twice_of_digit;
            }
            else
            {
                byte twice_of_digit_count = get_digit_count(twice_of_digit);
                long todm = 10;
                for (int j = 0; j < twice_of_digit_count; j++)
                {
                    total += get_digit(twice_of_digit, todm);
                    todm *= 10;
                }
            }
        }
        else
        {
            grand_total += digit;
        }
        m *= 10;
    }
    grand_total += total;

    if (grand_total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    long lastdigit = get_digit(credit_card_number, pow(10, credit_card_digit_count));
    long last2digit = get_digit(credit_card_number, pow(10, credit_card_digit_count - 1)) + (10 * lastdigit);

    if (lastdigit == VISA && credit_card_digit_count >= VISA_CARD_LEN && credit_card_digit_count <= VISA_CARD_LEN_2)
        printf("VISA\n");
    else if ((last2digit == MASTER_CARD || last2digit == MASTER_CARD_2 || last2digit == MASTER_CARD_3 || last2digit == MASTER_CARD_4 || last2digit == MASTER_CARD_5) && credit_card_digit_count == MASTER_CARD_LEN)
        printf("MASTERCARD\n");
    else if ((last2digit == AMERICAN_EXPRESS || last2digit == AMERICAN_EXPRESS_2) && credit_card_digit_count == AMERICAN_EXPRESS_LEN)
        printf("AMEX\n");
    else
        printf("INVALID\n");
    return 0;
}